### PR TITLE
feat(modern): auto detect modern mode

### PR DESCRIPTION
**Pull request by clarkdo**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

If user doesn't specify `modern` in `command` and `nuxt.config.js`, a middleware will detect by if there is modern built asset in dist:
1. ssr: server
1. spa: client
1. no modern assets: false

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

### CODE REVIEW

File: packages/server/src/middleware/modern.js [hunk 1]
```diff
@@ -11,12 +13,33 @@ const isModernBrowser = (ua) => {
...
 
-export default function (req, res, next) {
-  const { socket = {}, headers } = req
-  if (socket.modernMode === undefined) {
-    const ua = headers && headers['user-agent']
-    socket.modernMode = isModernBrowser(ua)
+let detected = false
+
+const detectModernBuild = ({ options, resources }) => {
+  if (detected === false && ![false, 'client', 'server'].includes(options.modern)) {
```
Comment (pi0) [inline packages/server/src/middleware/modern.js]: Can't we relax condition to `typeof options.modern !== `undefined`?

File: packages/server/src/middleware/modern.js [hunk 2]
```diff
@@ -11,12 +13,33 @@ const isModernBrowser = (ua) => {
...
+    }
+  }
+}
+
+const detectModernBrowser = (req, options) => {
+  if (options.modern === 'server') {
+    const { socket = {}, headers } = req
+    if (socket.modernMode === undefined) {
+      const ua = headers && headers['user-agent']
+      socket.modernMode = isModernBrowser(ua)
```
Comment (pi0) [inline packages/server/src/middleware/modern.js]: Why it is directly set on `socket` object?

**Review by pi0 — APPROVED** [review]

File: packages/server/src/middleware/modern.js (hunk 1 above)
Comment (clarkdo) [reply packages/server/src/middleware/modern.js]: I think what you mean is `modern === undefined`?

If user specified `modern:true`, we also need to detect the mode 😄

File: packages/server/src/middleware/modern.js (hunk 2 above)
Comment (clarkdo) [reply packages/server/src/middleware/modern.js]: For reducing the matching times, we only match one time in a connection.

Comment (pi0) [reply packages/server/src/middleware/modern.js]: Damn good idea 👍 It worth mentioning in the comments :D

**Comment by clarkdo** [comment]
Resolve #4422

**Pull request merged**